diff --git a/src/server/protocol.d.ts b/src/server/protocol.d.ts
index ef9ad79..fa2528f 100644
--- a/src/server/protocol.d.ts
+++ b/src/server/protocol.d.ts
@@ -870,6 +870,8 @@ declare module ts.server.protocol {
           * errors for the files in the file list
           */
         delay: number;
+
+        responseType?: string;
     }
 
     /**
diff --git a/src/server/session.ts b/src/server/session.ts
index a4cb3e5..4fdc00e 100644
--- a/src/server/session.ts
+++ b/src/server/session.ts
@@ -693,6 +693,29 @@ module ts.server {
             }
         }
 
+        getErrorResponse(fileNames: string[]) {
+            var checkList = fileNames.reduce((accum: PendingErrorCheck[], fileName: string) => {
+                fileName = ts.normalizePath(fileName);
+                var project = this.projectService.getProjectForFile(fileName);
+                if (project) {
+                    accum.push({ fileName, project });
+                }
+                return accum;
+            }, []);
+
+            return checkList.map(checkSpec => {
+                var project = checkSpec.project;
+                var file = checkSpec.fileName;
+                var semanticDiags = project.compilerService.languageService.getSemanticDiagnostics(file);
+                var syntaticDiags = project.compilerService.languageService.getSyntacticDiagnostics(file);
+                return {
+                    file: file,
+                    syntaxDiag: syntaticDiags ? syntaticDiags.map((diag) => formatDiag(file, project, diag)) : [],
+                    semanticDiag: semanticDiags ? semanticDiags.map((diag) => formatDiag(file, project, diag)) : [],
+                }
+            })
+        }
+
         change(line: number, offset: number, endLine: number, endOffset: number, insertString: string, fileName: string) {
             var file = ts.normalizePath(fileName);
             var project = this.projectService.getProjectForFile(file);
@@ -911,8 +934,12 @@ module ts.server {
                     }    
                     case CommandNames.Geterr: {
                         var geterrArgs = <protocol.GeterrRequestArgs>request.arguments;
-                        response = this.getDiagnostics(geterrArgs.delay, geterrArgs.files);
-                        responseRequired = false;
+                        if (geterrArgs.responseType === 'response') {
+                            response = this.getErrorResponse(geterrArgs.files);
+                        } else {
+                            response = this.getDiagnostics(geterrArgs.delay, geterrArgs.files);
+                            responseRequired = false;
+                        }
                         break;
                     }
                     case CommandNames.Change: {
